### PR TITLE
HOCS-4182 Expensive calls to GetStageUser GetStageTeam and GetStageType

### PR DESCRIPTION
**Pull request by nickmcmahon01**
This commit replaces the use of active_case to provide the original intent of the query as active_case has changed.

This change reduces the execution time from 500ms to 0.2ms

### CODE REVIEW

**Comment by nickmcmahon01** [comment]
I renamed the branch  and it killed the PR, see comments here;
https://github.com/UKHomeOffice/hocs-casework/pull/480

File: src/main/java/uk/gov/digital/ho/hocs/casework/api/StageService.java [hunk 1]
```diff
@@ -355,15 +355,22 @@ private Stage getStage(UUID caseUUID, UUID stageUUID) {
         }
     }
 
+    private BasicStage getBasicStage(UUID caseUUID, UUID stageUUID) {
```
Comment (fewtrees-ho) [inline src/main/java/uk/gov/digital/ho/hocs/casework/api/StageService.java:358]: Name needs to reflect its purpose. Perhaps renaming the other version is more appropriate i.e. `getStageTypeAndCaseFromStageData()` ?

File: src/main/java/uk/gov/digital/ho/hocs/casework/domain/model/BasicStage.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+import javax.persistence.Table;
+import java.time.LocalDateTime;
+import java.util.UUID;
+
+import static uk.gov.digital.ho.hocs.casework.application.LogEvent.STAGE_CREATE_FAILURE;
+
+@NoArgsConstructor
+@Entity
+@Table(name = "stage")
+public class BasicStage extends BaseStage {
```
Comment (fewtrees-ho) [inline src/main/java/uk/gov/digital/ho/hocs/casework/domain/model/BasicStage.java:16]: `BasicStage` vs `BaseStage` is confusing. Why not just `Stage`?

File: src/main/java/uk/gov/digital/ho/hocs/casework/domain/model/Stage.java [hunk 3]
```diff
@@ -195,4 +158,12 @@ public boolean isActive() {
         return this.teamUUID != null;
     }
 
+    public void update(Map<String, String> newData, ObjectMapper objectMapper) {
```
Comment (fewtrees-ho) [inline src/main/java/uk/gov/digital/ho/hocs/casework/domain/model/Stage.java:161]: These methods are already part of `AbstractJsonDataMap` so why here?

**Review by fewtrees-ho — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/digital/ho/hocs/casework/domain/model/BasicStage.java (hunk 2 above)
Comment (nickmcmahon01) [reply src/main/java/uk/gov/digital/ho/hocs/casework/domain/model/BasicStage.java:16]: I was going for the minimum diff for the PR, happy to rename the current 'Stage' to 'StageData' to reflect that it returns the data blob and 'BasicStage' to 'Stage' ? That would be more correct?

File: src/main/java/uk/gov/digital/ho/hocs/casework/api/StageService.java (hunk 1 above)
Comment (nickmcmahon01) [reply src/main/java/uk/gov/digital/ho/hocs/casework/api/StageService.java:358]: Good suggestion

File: src/main/java/uk/gov/digital/ho/hocs/casework/domain/model/Stage.java (hunk 3 above)
Comment (nickmcmahon01) [reply src/main/java/uk/gov/digital/ho/hocs/casework/domain/model/Stage.java:161]: We can't extend 2 classes and we need to extend BaseStage as a @MappedSuperclass. I was about to suggest we move AbstractJsonDataMap to a static 'helper' class but I've just noticed it already is 'JsonMapUtils', 'AbstractJsonDataMap' is just a 1 line method to call the helper class so it doesn't need to exist in my opinion - we never pass Stage and CaseData into the same method that uses that base class for example. I'll remove it from both and call the helper class directly...quite why we need a helper class to use a map is for another day 😄

File: src/main/java/uk/gov/digital/ho/hocs/casework/domain/model/BasicStage.java (hunk 2 above)
Comment (fewtrees-ho) [reply src/main/java/uk/gov/digital/ho/hocs/casework/domain/model/BasicStage.java]: Or `StageComposite` would suggest there's more in it
`StageWithCase` - is explicit and is without confusion..

`BasicStage` to `Stage` is correct.

**Review by fewtrees-ho — APPROVED** [review]
Thanks

**Pull request merged**